### ISSUE TITLE
Implementing a "blacklist" in sympy.

### ISSUE THREAD

**Issue by Kadelka** [sympy/sympy#26350]
Is it possible to define some sort of blacklist in sympy? Problem: I have a class, say `class Octionion(Quaternion)` which uses sympy.Quaternion. In my program I have

`u = Quaternion(2,3)*a`

where `a` an instance of Octonion. Since Quaternion has the method `__mul__(self,a)` the interpreter, maybe python, calls this method and the result is `SympifyError: SympifyError: octonion(a0, a1, a2, a3, a4, a5, a6, a7)`, which unfortunately seems to be not recoverable. `u = Octonion(2,3)*a` works, similarly `u = a*Quaternion(2,3)`. Here `__mul__(self,Quaternion(2,3))` with `__mul__` from Octonion is used, which automatically changes `Quaternion(2,3)` to `Octonion(2,3)`. Note that multiplication is not commutative, so that the last variant cannot be used. The same problem exists for other operators.

One idea is that a "blacklist" is implemented in Quaternion such that a call like `__mul__(Quaternion,unknown)` immediately returns NotImplemented with the consequence that now `__rmul__(unknown,Quaternion)` returns the correct value. Is this possible or is there a workaround?

**Comment by Kadelka** [comment]
Meanwhile I found monkey patching as a workaround for my problem. The following code works in my module octonion:

`from sympy import Quaternion, core

### Some code with class Octonion():

Quaternion.Blacklist = []
Quaternion.Blacklist.append(Octonion)

def dummy_add(self,other):
    if type(other) in Quaternion.Blacklist:
        return NotImplemented
    else:
        return self.add(other)
Quaternion.__add__ = dummy_add

def dummy_sub(self,other):
    if type(other) in Quaternion.Blacklist:
        return NotImplemented
    else:
        return self.add(other*-1)
Quaternion.__sub__ = dummy_sub

def dummy_mul(self,other):
    if type(other) in Quaternion.Blacklist:
        return NotImplemented
    else:
        return self._generic_mul(self, core.sympify(other,strict=True))
Quaternion.__mul__ = dummy_mul

def dummy_truediv(self,other):
    if type(other) in Quaternion.Blacklist:
        return NotImplemented
    else:
        return self * core.sympify(other)**-1
Quaternion.__truediv__ = dummy_truediv
`

This works as needed and doesn't change the behaviour  of `Quaternion`. I think this code can be incorporated in quaternion.py (of course without `Quaternion.Blacklist.append(Octonion)`). Only `dummy_*` has to be replaced by `__*__`.

**Comment by oscarbenjamin** [comment]
We don't need a blacklist. Just Quaternion should catch SympifyError and return NotImplemented.

**Comment by Kadelka** [comment]
I agree in the long run. But I think that without Blacklists the logic of sympy has to be changed. It will be difficult to implement, since we have to cope with totally unknown classes.

**Comment by sylee957** [comment]
Where does your `Octonion` come from?
https://github.com/search?q=repo%3Asympy%2Fsympy%20Octonion&type=code

I think that your problem is that you were trying to implement your custom class `Octonion` and face some errors, so I think that
you should subclass `Octonion` from `Basic` or `Expr`.

**Comment by asmeurer** [comment]
> Is there a procedure in sympy which makes it possible to discern between allowed objects and unknown objects?

Allowed objects are those that can be sympified into Basic instance. If you want your `octonion` object to be part of SymPy expression trees, it needs to be a `Basic` subclass. Otherwise, it should at least be converted to one with `sympify` (see the "extending" section [here](https://docs.sympy.org/latest/modules/core.html#module-sympy.core.sympify)).

If your class is subclassing Quaternion, it should already work here, so it sounds like there might just be a bug in the Quaternion `__mul__` implementation.

**Comment by asmeurer** [comment]
By the way, don't confuse the commutativity of multiplication with `__mul__` vs. `__rmul__`. `x*y` is the same as `x.__mul__(y)` or `y.__rmul__(x)`. Neither is the same as `y*x`. `__rmul__` (and the other `r` operators) exists to allow objects to define their behavior when they are on the right-hand side of the `*` operator.

**Comment by Kadelka** [comment]
Actually for my project there is a great difference between `__mul__` and `__rmul__`. Actually these are of the form `def __mul__(self,v)` and `def __rmul__(self,v)'. Which form is used depends on `self`. Python only uses `__rmul__` if `__mul__` return with a `NotImplemented`. And this is my problem. I don't want Octonion and Sedenion to be integrated in sympy (I think it's difficult). I simply want to use `Quaternion` like complex or real numbers. I like the class Octonion. It even allows that the components of `Octonion` and `Sedenion` are complex numbers like `1+2*I`.

**Comment by asmeurer** [comment]
I think what you're looking for could be done with `_op_priority`, except I think Quaternion, in implementing its own `__mul__` logic, bypasses it. Probably Quaternion should be using the dispatch mechanisms on Mul, similar to the matrix expressions.

**Comment by oscarbenjamin** [comment]
The problem here is just that `Quaternion.__mul__` should be changed to return `NotImplemented` for unrecognised types. Then `Octonion.__mul__` can handle multiplication between octonions and quaternions.

**Comment by Kadelka** [comment]
The problem actually is: **When** should `Quaternion.__mul__` return `NotImplemented`. A "whitelist" seems to be too complicated and isn't fast. A blacklist, which is simple, seems to be not desired.

**Comment by oscarbenjamin** [comment]
No, methods like `__mul__` should always use a white list: check for types that are allowed and otherwise return `NotImplemented` so that the other operand's `__rmul__` can be called. This has always been how Python's cooperative dispatch method works but few people seem to understand this and as a result many people write buggy implementations of these methods. A method like `__mul__` should never raise `TypeError` or raise an exception but just return `NotImplemented` for unrecognised types. If `__mul__` and `__rmul__` both return `NotImplemented` then the Python run-time will raise `TypeError`.

**Comment by sylee957** [comment]
I have suggestion that maybe here can be changed to return NotImplemented
I'm busy quite now, so I have not concluded some experiments, however, 

https://github.com/sympy/sympy/blob/9d1306140f034f62c69bce4c2da8da1a9b771d0d/sympy/algebras/quaternion.py#L749-L750

I generally suggest to return `NotImplemented` as much possible, in Python, and not to raise any errors as much as possible in `__add__, __mul__` (maybe except for things like zero divisions). That can be some antipattern, and we can encounter extensibility issues like your issue.

I'm also trying to find good reference material for, when should it return NotImplemented, and I find this resources. It may not have strong opinion about it, however, it may be reference pattern to follow

https://docs.python.org/3/library/numbers.html#implementing-the-arithmetic-operations

I may also question, like if using `sympify` is valid there because I'm aware that it throws errors for unknown types too, which SymPy objects may better just return `NotImplemented` and do not eliminate the possibility for foreign objects to handle it.

> A "whitelist" seems to be too complicated and isn't fast. A blacklist, which is simple, seems to be not desired. A possible simple solition may be my proposal but with hidden blacklist.

Unfortunately, I've not heard about what is blacklist and whitelist, and if that semantically makes sense for very simple add, mul Quaternion, or Octonion, and whether something is performant or simple.

However, I understand that what Python arithmetic overrides and `NotImplemented` does, 
sounds something similar like pattern matching, which is more vanilla feature of Python, and generally generalizes blacklist or whitelist. So we should try to use that, than introducing some new machinery.
